scripts/digest_traces.py

```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python3
+""" This script processes a traces file: it extracts the requests that seem
+    to be consuming memory.
+
+    ALS.IN and ALS.OUT need to be active for this to work.
+
+    Usage:
+
+        digest_traces.py <traces_file>
+"""
+
+import json
+import os
+import re
+import sys
+
+TRACES_LINE = \
+  ' *\\[ALS\\.(IN|OUT)\\] ({.*}) \\[Watermark:. (\\d+)/ (\\d+)\\].*'
+# Regexp to parse one line of traces
+
+
+class Digester(object):
+    """ The class that takes care of the file parsing and counting """
+
+    def __init__(self, filename):
+        """Initialize. filename is the path to the traces file"""
+
+        self.in_re = re.compile(TRACES_LINE)
+
+        self.last_method = "(elaboration)"
+        self.last_memory_usage = 0
+        # the last method and last memory usage received
+
+        self.methods_to_memory = {}
+        # cumulative memory increase indexed by method id
+
+        self.methods_exception_count = {"(elaboration)": 0}
+        # count the exception occurrences for each exception
+
+        self.currently_looking_at_exception = False
+
+        self.methods_count = {"(elaboration)": 1}
+        # methods count indexed by method id
+
+        # read the file
+        with open(filename, "r") as f:
+            for line in f.readlines():
+                self.read_line(line.strip())
+
+    def read_line(self, line):
+        """Read and process one line of traces"""
+        if (not self.currently_looking_at_exception and
+           line.startswith("_ALS.MAIN")):
```

Comment:
You are assuming that any multi-lines trace in "ALS.MAIN" is an exception.

Reply:
yes, in this version we have no better way of doing this! I will modify the ALS to say "EXCEPTION" shortly.